### PR TITLE
feat: thorchain memo support

### PR DESCRIPTION
**Pull request by elmutt**
Add a function to return thorchain memo shortened to a max of 80 characters as described:

https://dev.thorchain.org/thorchain-dev/memos

### CODE REVIEW

**Review by DaoDev44 — DISMISSED** [review]

File: packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.test.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { makeSwapMemo } from './makeSwapMemo'
+
+describe('makeSwapMemo', () => {
+  it('make a trade to usdc memo with', () => {
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.test.ts:4]: nitpick that test descriptions start with `should` and this one isn't a complete sentence

File: packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import { adapters } from '@shapeshiftoss/caip'
+
+import { SwapError, SwapErrorTypes } from '../../../api'
+
+// BTC (and likely other utxo coins) can only support up to 80 character memos
+const MAX_LENGTH = 80
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts:6]: We are going to need to think about how to do a reverse fuzzy lookup on abbreviated assets in order to turn it into a caip assetId if there is not a ticket for that already

Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts:6]: oh thats a fun one

File: packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * @returns thorchain memo shortened to a max of 80 characters as described:
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts:9]: better link to show memo message limits
https://dev.thorchain.org/thorchain-dev/memos#mechanism-for-transaction-intent

File: packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  if (fullMemo.length <= MAX_LENGTH) return fullMemo
+
+  const fullMemoLength = fullMemo.length
+  const truncateAmount = fullMemoLength - MAX_LENGTH
+  const thorIdSymbol = thorId.slice(0, thorId.indexOf('-'))
+  const thorIdReferenceTruncated = thorId.slice(thorId.length - truncateAmount, thorId.length)
+  const truncatedThorId = `${thorIdSymbol}-${thorIdReferenceTruncated}`
+  const shortenedMemo = `s:${truncatedThorId}:${destinationAddress}:${limit}`
+  return shortenedMemo
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts:42]: suggestion: thorchain verbiage, comment on abbreviation, catch if we don't have an asset to abbreviate but the memo is too long (not sure how this case would be handled?), and one liner to abbreviate
```suggestion
  const memo = `s:${thorId}:${destinationAddress}:${limit}`

  if (memo.length <= MAX_LENGTH) return memo

  const abbreviationAmount = memo.length - MAX_LENGTH

  // delimeter between ticker and id allowing us to abbreviate the id: https://dev.thorchain.org/thorchain-dev/memos#asset-notation
  const delimeterIndex = memo.indexOf('-')

  if (delimeterIndex === -1) {
    throw new SwapError('[makeSwapMemo] - unable to abbreviate asset, no delimeter found', {
      code: SwapErrorTypes.MAKE_MEMO_FAILED
    })
  }

  return memo.replace(memo.slice(delimeterIndex + 1, delimeterIndex + abbreviationAmount), '')
```

File: packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts (hunk 2 above)
Comment (0xdef1cafe) [reply packages/swapper/src/swappers/thorchain/utils/makeSwapMemo.ts:6]: @elmutt i agree with @kaladinlight  and that logic absolutely needs to go into the caip adapter

**Review by DaoDev44 — APPROVED** [review]

**Pull request merged**